src/runtime-rs/src/shim/src/config.rs

```diff
@@ -57,14 +52,39 @@ fn get_default_config_path() -> Option<PathBuf> {
     None
 }
 
-pub fn load_configuration() -> ConfigResult<TomlConfig> {
-    let path = get_preferred_config_path()
+/// Try to load configuration information from a configuration file.
+///
+/// Try to load configuration information from a configuration file in following order:
+/// - configuration file specified by config.json from OCI image bundle
+/// - configuration file specified by the "KATA_CONF_FILE" environment variable
+/// - /etc/kata-containers/configuration.toml
+/// - /usr/share/defaults/kata-containers/configuration.toml
+pub fn load_configuration(bundle_path: &PathBuf) -> Result<TomlConfig, ConfigError> {
+    let path = get_preferred_config_path(bundle_path)
         .or_else(get_default_config_path)
         .ok_or(ConfigError::Get)?;
+    let content = std::fs::read_to_string(path).map_err(ConfigError::Load)?;
+
+    toml::from_str(&content).map_err(ConfigError::Parse)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use test_utils::{get_image_bundle_path, get_kata_config_file};
 
-    let contents = std::fs::read_to_string(path).map_err(ConfigError::Load)?;
+    #[test]
+    fn test_load_from_env() {
+        let file = get_kata_config_file();
+        std::env::set_var("KATA_CONF_FILE", file);
+        let bundle_path = get_image_bundle_path();
+        let config = load_configuration(&bundle_path).unwrap();
+        assert_eq!(config.runtime.debug, true);
 
-    let config = toml::from_str(&contents).map_err(ConfigError::Parse)?;
+        let dragonball = config.hypervisor.get("dragonball").unwrap();
+        assert_eq!(dragonball.default_vcpus, 2);
 
-    Ok(config)
+        let qemu = config.hypervisor.get("qemu").unwrap();
+        assert_eq!(qemu.default_vcpus, 4);
+    }
 }
```

Comment:
All the changes above here look like they're probably worthwhile.  However, they have nothing to do with the `ShimExecutor` rework.  This is what I mean by the commits not being logically separated.

Reply:
Fixed in the second revision: rebased the patches by logical separations